### PR TITLE
feat: Add support for all Prometheus metric types + internal refactor.

### PR DESCRIPTION
**Pull request by bwplotka**
This PR refreshes avalanche binary, while adding support for all Prometheus metric type.

This should not break compatibility (even if we can), but I deprecate one flag.

Fixes https://github.com/prometheus-community/avalanche/issues/77

High level changes (as in the CHANGLOG):
* add support for all metric types
* deprecated --metric-count flag
*  --*-interval flags set to 0 means no change logic
* added OpenMetrics support

Refactor:
* Unglobalized flags and registry/metrics.
* Moved trillion arguments to one config struct for metrics; moved flags next to config.
* Use unchecked collector instead of unregistering metrics all the time
* Add sigterm/interrupt catching
* Updated tests

### CODE REVIEW

File: README.md
Comment (saswatamcode) [inline README.md:8]:
```suggestion
This allows load testing services that can scrape (e.g. Prometheus, OpenTelemetry Collector and so), as well as, services accepting data via Prometheus remote_write API such as [Thanos](https://github.com/thanos-io/thanos), [Cortex](https://github.com/cortexproject/cortex), [M3DB](https://m3db.github.io/m3/integrations/prometheus/), [VictoriaMetrics](https://github.com/VictoriaMetrics/VictoriaMetrics/) and other services [listed here](https://prometheus.io/docs/operating/integrations/#remote-endpoints-and-storage).
```

File: metrics/serve.go
Comment (saswatamcode) [inline metrics/serve.go:198]: Probably don't need the comment, given help text
```suggestion
```

File: metrics/write.go
Comment (saswatamcode) [inline metrics/write.go:111]: Maybe refactor for later, but the write method fails after 20 remote write errors i.e 20 ticks, I wonder if we can just fail early, or log each error as they happen, as your requestInterval might be large and you might need to wait.

File: cmd/avalanche.go
Comment (saswatamcode) [inline cmd/avalanche.go:76]: Might make sense to do similar NewConfigFromFlag method for remote write flags?
We already have ConfigWrite struct

File: metrics/serve.go
Comment (saswatamcode) [inline metrics/serve.go:269]: Nice use of generics!

**Review by saswatamcode — APPROVED** [review]
Thanks a lot for doing this refactor and adding all types! Looks really neat! ❤️ 
Some small comments.

File: cmd/avalanche.go
Comment (bwplotka) [reply cmd/avalanche.go:76]: Definitely, but not in this PR 🤗

File: metrics/write.go
Comment (bwplotka) [reply metrics/write.go:112]: Do you mind adding an issue? I didn't mean to change client here

Comment (saswatamcode) [reply metrics/write.go:112]: Sure will add!

**Review by saswatamcode — APPROVED** [review]
LGTM!

File: metrics/write.go
Comment (saswatamcode) [reply metrics/write.go:112]: https://github.com/prometheus-community/avalanche/issues/81

**Review by ArthurSens — APPROVED** [review]
Big PR to review without much involvement in the project hehe

I've tried it by running and manually checking the /metrics endpoint, LGTM! This looks perfect for benchmarking our work with @Maniktherana in prometheus/prometheus. And if we implement PRWv2 it also looks perfect for the work we're doing with @jmichalek132 in the collector :)

File: cmd/avalanche.go
Comment (bwplotka) [reply cmd/avalanche.go:166]: 👍🏽 Perhaps behind flag, but yes

File: metrics/serve.go
Comment (bwplotka) [reply metrics/serve.go:99]: We can remove yes, just didn't want to introduce too many changes at once -- this might be the MOST used flag (most important one), so wanted to give ppl time to adapt.

This also reminded me I might want to walk through docs to update them if they use metric-count.

Let's take it easy - added issue about this https://github.com/prometheus-community/avalanche/issues/82

**Comment by beorn7** [comment]
Is there anything in particular I should review here? I don't think I will find time for a full review anytime soon. I don't even know this project.

**Comment by bwplotka** [comment]
> Is there anything in particular I should review here? I don't think I will find time for a full review anytime soon. I don't even know this project.

Yea don't worry - I didn't know this codebase yesterday either. But anyway, I am actually no longer sure why I pinged you, sorry for spam and considering 🙈 

Since no objections I will allow myself to merge this, we can always revert. I  pinged current maintainers and not sure if they are still active in this project, will try more and perhaps propose myself in the meantime here.

**Pull request merged**
